presto-native-execution/presto_cpp/main/tests/LinuxMemoryCheckerTest.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "presto_cpp/main/LinuxMemoryChecker.cpp"
+#include <boost/algorithm/string.hpp>
+#include <boost/filesystem.hpp>
+#include <gtest/gtest.h>
+#include "velox/common/base/VeloxException.h"
+#include "velox/common/base/tests/GTestUtils.h"
+
+namespace fs = boost::filesystem;
+
+namespace {
+std::string getStatsFilePath(const std::string& fileName) {
```

Comment:
This seems unnecessary. Look at how Parquet files are loaded in Velox tests.

Reply:
In Velox, [ParquetTableScanTest.cpp](https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/tests/reader/ParquetTableScanTest.cpp) uses getDataFilePath() from [DataFiles.cpp](https://github.com/facebookincubator/velox/blob/main/velox/dwio/common/tests/utils/DataFiles.cpp).

```
std::string getDataFilePath(
    const std::string& baseDir,
    const std::string& filePath) {
  std::string current_path = fs::current_path().c_str();
  if (boost::algorithm::ends_with(current_path, "fbcode")) {
    return current_path + "/" + baseDir + "/" + filePath;
  }
  return current_path + "/" + filePath;
}
```

This implementation doesn't work for my case due to the cwd being in /presto/presto-native-execution so I would have to append "/presto_cpp/main/tests/examples/" which is why I made my own implementation here.

I also see [AnnouncerTest.cpp](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/tests/AnnouncerTest.cpp#L28) also making its own implementation to get the certs path. I'm not sure how it's able to find the certs folder since it would need to append "/presto_cpp/main/tests/certs/"...

```
std::string getCertsPath(const std::string& fileName) {
  std::string currentPath = fs::current_path().c_str();
  if (boost::algorithm::ends_with(currentPath, "fbcode")) {
    return currentPath +
        "/github/presto-trunk/presto-native-execution/presto_cpp/main/tests/certs/" +
        fileName;
  }

  // CLion runs the tests from cmake-build-release/ or cmake-build-debug/
  // directory. Hard-coded json files are not copied there and test fails with
  // file not found. Fixing the path so that we can trigger these tests from
  // CLion.
  boost::algorithm::replace_all(currentPath, "cmake-build-release/", "");
  boost::algorithm::replace_all(currentPath, "cmake-build-debug/", "");

  return currentPath + "/certs/" + fileName;
}
```

Reply:
Trying out this implementation:
```
std::string getStatsFilePath(const std::string& fileName) {
  std::string currentPath = fs::current_path().c_str();
  if (boost::algorithm::ends_with(currentPath, "fbcode")) {
    return currentPath +
        "/github/presto-trunk/presto-native-execution/presto_cpp/main/tests/examples/" +
        fileName;
  }

  return currentPath + "/examples/" + fileName;
}
```

It looks like it depends on where I'm running the test executable from:
```
root@biday3ember11:~/presto/presto-native-execution# /root/presto/presto-native-execution/_build/debug/presto_cpp/main/tests/presto_server_test --gtest_filter=LinuxMemoryCheckerTest.memoryStatFileV1
Note: Google Test filter = LinuxMemoryCheckerTest.memoryStatFileV1
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LinuxMemoryCheckerTest
[ RUN      ] LinuxMemoryCheckerTest.memoryStatFileV1
unknown file: Failure
C++ exception with description "open("/root/presto/presto-native-execution/examples/cgroupV1memory.stat", 0, 00666) failed: No such file or directory" thrown in the test body.
[  FAILED  ] LinuxMemoryCheckerTest.memoryStatFileV1 (1 ms)
[----------] 1 test from LinuxMemoryCheckerTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LinuxMemoryCheckerTest.memoryStatFileV1

 1 FAILED TEST
 ```
 
 ```
root@biday3ember11:~/presto/presto-native-execution# cd presto_cpp/main/tests
root@biday3ember11:~/presto/presto-native-execution/presto_cpp/main/tests# /root/presto/presto-native-execution/_build/debug/presto_cpp/main/tests/presto_server_test --gtest_filter=LinuxMemoryCheckerTest.memoryStatFileV1
Note: Google Test filter = LinuxMemoryCheckerTest.memoryStatFileV1
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LinuxMemoryCheckerTest
[ RUN      ] LinuxMemoryCheckerTest.memoryStatFileV1
[       OK ] LinuxMemoryCheckerTest.memoryStatFileV1 (0 ms)
[----------] 1 test from LinuxMemoryCheckerTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
```

Reply:
@majetideepak 

I added the following to /presto-native-execution/presto_cpp/main/tests/CMakeLists.txt and it worked:
```
if(DEFINED PRESTO_MEMORY_CHECKER_TYPE AND PRESTO_MEMORY_CHECKER_TYPE STREQUAL
                                          "LINUX_MEMORY_CHECKER")
  target_sources(presto_server_test PRIVATE LinuxMemoryCheckerTest.cpp)
  target_compile_definitions(presto_server_test PRIVATE TEST_WORKING_DIR="${CMAKE_CURRENT_SOURCE_DIR}")
endif()
```

Then I used TEST_WORKING_DIR in LinuxMemoryCheckerTest.cpp and it worked:
```
std::string getStatsFilePath(const std::string& fileName) {
  std::string currentPath = TEST_WORKING_DIR;
  if (boost::algorithm::ends_with(currentPath, "fbcode")) {
    return currentPath +
        "/github/presto-trunk/presto-native-execution/presto_cpp/main/tests/examples/" +
        fileName;
  }

  return currentPath + "/examples/" + fileName;
}
```

What do you think of this approach?

Reply:
You can skip the fbcode check.  Meta does not run this test.
We have already set the test working directory to this. See below.
The CMake expectation is that you run the ctest command and not run it manually as you did above.
IDE such as CLion also use ctest I believe. You should not see any problem there.

```
add_test(
  NAME presto_server_test
  COMMAND presto_server_test
  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
```